Mathlib/NumberTheory/Transcendental/LindemannWeierstrass.lean

```diff
@@ -0,0 +1,1441 @@
+    simp_rw [Algebra.algebraMap_eq_smul_one]
+    change (x • Finsupp.single 0 (1 : K s)) a = _
+    simp_rw [Finsupp.smul_single, Finsupp.single_apply]
+    split_ifs <;> rfl
+#align rat_coeff_equiv.aux RatCoeffEquiv.aux
+
+def ratCoeffEquiv : ratCoeff s ≃ₐ[ℚ] AddMonoidAlgebra ℚ (K s) :=
+  (RatCoeffEquiv.aux s).trans
+    (AddMonoidAlgebra.algEquivCongrLeft (IntermediateField.botEquiv ℚ (K s)))
+#align rat_coeff_equiv ratCoeffEquiv
+
+theorem ratCoeffEquiv_apply_apply (x : ratCoeff s) (i : K s) :
+    ratCoeffEquiv s x i =
+      (IntermediateField.botEquiv ℚ (K s)) ⟨(x : AddMonoidAlgebra (K s) (K s)) i, x.2 i⟩ :=
+  rfl
+#align rat_coeff_equiv_apply_apply ratCoeffEquiv_apply_apply
+
+theorem support_ratCoeffEquiv (x : ratCoeff s) :
+    (ratCoeffEquiv s x).support = (x : AddMonoidAlgebra (K s) (K s)).support := by
+  dsimp [ratCoeffEquiv, RatCoeffEquiv.aux]
+  rw [Finsupp.support_mapRange_of_injective]
+  exact AlgEquiv.injective _
+#align support_rat_coeff_equiv support_ratCoeffEquiv
+
+section
+
+variable (F : Type _) [Field F] [Algebra ℚ F]
+
+noncomputable def mapDomainFixed : Subalgebra F (AddMonoidAlgebra F (K s)) where
+  carrier := {x | ∀ f : Gal s, AddMonoidAlgebra.mapDomainAlgAut ℚ _ f.toAddEquiv x = x}
+  mul_mem' {a b} ha hb f := by rw [map_mul, ha, hb]
+  add_mem' {a b} ha hb f := by rw [map_add, ha, hb]
+  algebraMap_mem' r f := by
+    change Finsupp.equivMapDomain f.toEquiv (Finsupp.single _ _) = Finsupp.single _ _
+    rw [Finsupp.equivMapDomain_single]
+    change Finsupp.single (f 0) _ = _; rw [AlgEquiv.map_zero]
+#align map_domain_fixed mapDomainFixed
+
+instance : CoeFun (mapDomainFixed s F) fun _ => (K s) → F where
+  coe f := f.1
+
+theorem mem_mapDomainFixed_iff (x : AddMonoidAlgebra F (K s)) :
+    x ∈ mapDomainFixed s F ↔ ∀ i j, i ∈ MulAction.orbit (Gal s) j → x i = x j := by
+  simp_rw [MulAction.mem_orbit_iff]
+  change (∀ f : Gal s, Finsupp.equivMapDomain (↑(AlgEquiv.toAddEquiv f)) x = x) ↔ _
+  refine' ⟨fun h i j hij => _, fun h f => _⟩
+  · obtain ⟨f, rfl⟩ := hij
+    rw [AlgEquiv.smul_def, ← FunLike.congr_fun (h f) (f j)]
+    change x (f.symm (f j)) = _; rw [AlgEquiv.symm_apply_apply]
+  · ext i; change x (f.symm i) = x i
+    refine' (h i ((AlgEquiv.symm f) i) ⟨f, _⟩).symm
+    rw [AlgEquiv.smul_def, AlgEquiv.apply_symm_apply]
+#align mem_map_domain_fixed_iff mem_mapDomainFixed_iff
+
+noncomputable def mapDomainFixedEquivSubtype :
+    mapDomainFixed s F ≃
+      { f : AddMonoidAlgebra F (K s) // MulAction.orbitRel (Gal s) (K s) ≤ Setoid.ker f }
+    where
+  toFun f := ⟨f, (mem_mapDomainFixed_iff s F f).mp f.2⟩
+  invFun f := ⟨f, (mem_mapDomainFixed_iff s F f).mpr f.2⟩
+  left_inv _ := rfl
+  right_inv _ := rfl
+#align map_domain_fixed_equiv_subtype mapDomainFixedEquivSubtype
+
+end
+
+section toConjEquiv
+
+variable (F : Type _) [Field F] [Algebra ℚ F]
+
+open GalConjClasses
+
+def toConjEquiv : mapDomainFixed s F ≃ (GalConjClasses ℚ (K s) →₀ F) := by
```

Comment:
The LHS can be replaced by the subset of `A →₀ F` consisting of elements `x` such that `x a = x b` whenever `a ~ b` for some equivalence relation `~` on `A`, and `GalConjClasses ℚ (K s)` would be replaced by the quotient by the equivalence relation. This works for `Pi` always, but for `Finsupp` the right-to-left direction needs the equivalence classes to be finite.

The next 150+ lines seem to be mainly using this `Equiv` to transfer algebraic instances on the LHS to the RHS, and it's not yet clear to me why this is necessary; could we not just use this Equiv to map to the LHS and do all algebraic operations there?